File: 13-Stack-and-Queue/java/01-ImplementStackUsingArrays.java
class ArrayStack {
    private int[] stack;
    private int size;
    private static final int INITIAL_CAPACITY = 16;

    public ArrayStack() {
        stack = new int[INITIAL_CAPACITY];
        size = 0;
    }

    public void push(int x) {
        if (size == stack.length) {
            int[] newStack = new int[stack.length * 2];
            System.arraycopy(stack, 0, newStack, 0, size);
            stack = newStack;
        }
        stack[size++] = x;
    }

    public int pop() {
        return stack[--size];
    }

    public int top() {
        return stack[size - 1];
    }

    public boolean isEmpty() {
        return size == 0;
    }
}

File: 13-Stack-and-Queue/cpp/01-implement-stack-using-arrays.cpp
class ArrayStack {
private:
    vector<int> stack;

public:
    ArrayStack() {}

    void push(int x) {
        stack.push_back(x);
    }

    int pop() {
        int val = stack.back();
        stack.pop_back();
        return val;
    }

    int top() {
        return stack.back();
    }

    bool isEmpty() {
        return stack.empty();
    }
};
